test/core/tsi/ssl_transport_security_util_test.cc

```diff
@@ -0,0 +1,353 @@
+  // The first three bytes are always 0x17, 0x03, 0x03.
+  EXPECT_EQ(protected_output_frames[0], '\x17');
+  EXPECT_EQ(protected_output_frames[1], '\x03');
+  EXPECT_EQ(protected_output_frames[2], '\x03');
+  // The next two bytes are the size of the record, which is 5 bytes less
+  // than the size of the whole frame.
+  EXPECT_EQ(CalculateRecordSizeFromHeader(protected_output_frames[3],
+                                          protected_output_frames[4]),
+            expected_encrypted_bytes_size - 5);
+
+  std::vector<uint8_t> unprotected_output_bytes(plaintext_size);
+  std::size_t unprotected_output_bytes_size = unprotected_output_bytes.size();
+
+  // This frame should be decrypted by peer correctly.
+  EXPECT_EQ(ssl_protector_unprotect_util(
+                protected_output_frames.data(), peer_ssl, peer_bio,
+                &protected_output_frames_size, unprotected_output_bytes.data(),
+                &unprotected_output_bytes_size),
+            tsi_result::TSI_OK);
+  EXPECT_EQ(unprotected_output_bytes_size, unprotected_bytes_size);
+  unprotected_output_bytes.resize(unprotected_output_bytes_size);
+  unprotected_bytes.resize(unprotected_bytes_size);
+  EXPECT_THAT(unprotected_output_bytes, ContainerEq(unprotected_bytes));
+}
+
+class FlowTest : public TestWithParam<FrameProtectorUtilTestData> {
+ protected:
+  void SetUp() override {
+    ASSERT_EQ(DoHandshake(&client_ssl, &server_ssl), absl::OkStatus());
+
+    ASSERT_THAT(client_ssl, NotNull());
+    ASSERT_THAT(server_ssl, NotNull());
+
+    BIO* client_ssl_bio = nullptr;
+    ASSERT_EQ(BIO_new_bio_pair(&client_bio, /*writebuf1=*/0, &client_ssl_bio,
+                               /*writebuf2=*/0),
+              1);
+    SSL_set_bio(client_ssl, client_ssl_bio, client_ssl_bio);
+
+    BIO* server_ssl_bio = nullptr;
+    ASSERT_EQ(BIO_new_bio_pair(&server_bio, /*writebuf1=*/0, &server_ssl_bio,
+                               /*writebuf2=*/0),
+              1);
+    SSL_set_bio(server_ssl, server_ssl_bio, server_ssl_bio);
+
+    client_buffer_offset = 0;
+    client_buffer.resize(kMaxPlaintextBytesPerTlsRecord);
+    server_buffer_offset = 0;
+    server_buffer.resize(kMaxPlaintextBytesPerTlsRecord);
+  }
+
+  void TearDown() override {
+    BIO_free(client_bio);
+    SSL_free(client_ssl);
+    BIO_free(server_bio);
+    SSL_free(server_ssl);
+  }
+
+  SSL* client_ssl;
+  BIO* client_bio;
+  std::vector<uint8_t> client_buffer;
+  std::size_t client_buffer_offset;
+
+  SSL* server_ssl;
+  BIO* server_bio;
+  std::vector<uint8_t> server_buffer;
+  std::size_t server_buffer_offset;
+};
+
+TEST_P(FlowTest, TestWorkFlows) {
+  SslFrameProtectorUtilFunctionTestHelper(
+      GetParam().plaintext_size, GetParam().expected_encrypted_bytes_size,
+      client_ssl, client_bio, client_buffer, client_buffer_offset, server_ssl,
+      server_bio, server_buffer, server_buffer_offset);
+  SslFrameProtectorUtilFunctionTestHelper(
```

Comment:
I feel like it could be better if we can have something like:

```
TEST_P(TsiFrameProtectorTest, ClientSslBufferTests) {
// ...
}
TEST_P(TsiFrameProtectorTest, ServerSslBufferTests) {
// ...
}
```
So that we can get rid of those helper functions.

Reply:
I don't know think it is possible to do that. Rationale: all functions here are for the purpose of encrypting the message from one side so that the other side can decrypt it. So it is impossible to test on one side (Client and Server). What we are testing are two directions: client -> server and server-> client.

Reply:
Sorry I wasn't clear. I wasn't suggesting to change the code logic; I was just recommending another way to structure things out.
The idea behind this is that I am generally not a fan of "XXXTestHelper". In my opinion, it doesn't contain much information, and readers still have to look into that helper to understand what it really does. What I was suggesting was something like 

```
TEST_P(TsiFrameProtectorTest, ClientToServerBufferTests) { 
  // ... some preparation
  EXPECT_EQ(ssl_protector_protect_util(...), tsi_result::TSI_OK);
  // check buffer size&content, etc, this can be a static function shared by two test functions
  CheckBufferAfterProtection(...);
  EXPECT_EQ(ssl_protector_unprotect_util(...), tsi_result::TSI_OK);
  // check buffer size&content, etc, this can be a static function shared by two test functions
  CheckBufferAfterUnprotection(...);
} 

TEST_P(TsiFrameProtectorTest, ServerToClientBufferTests) { 
  // similar as above 
}
```
That way, it's clear for the reader that only the part `ssl_protector_protect_util` and `ssl_protector_unprotect_util` changed between two tests, and it's clear what these two tests are for.

Reply:
Changed the test structure as suggested.